### PR TITLE
State paths with regex

### PR DESCRIPTION
**Pull request by idantene**
- Complements endpoints with matching regex pattern
  - at the moment this is very naïve - does not check against the provided `schema.type`, but rather accepts everything and provides a named regex matching for later retrieval.
- Moves method/endpoint checks to `Service` instead of `ServiceStore`
- Adds `IService` interface
- Adds tests
- Modifies tests for some edge cases (i.e. when there are no `paths` - it now throws an error; what does it mean to set a state for a service that cannot be accessed? perhaps we should even throw when trying to create such a service to begin with, considering it an invalid schema?)

### CODE REVIEW

File: packages/unmock-core/src/__tests__/generator.test.ts [hunk 1]
```diff
@@ -1,7 +1,8 @@
 import fs from "fs";
 import yaml from "js-yaml";
 import { mockGeneratorFactory } from "../generator";
-import { ISerializedRequest, OASSchema, RequestToSpec } from "../interfaces";
+import { ISerializedRequest, RequestToSpec } from "../interfaces";
```
Comment (idantene) [inline packages/unmock-core/src/__tests__/generator.test.ts]: When we'll get to a more generator-specific PR, these would probably change (i.e. `RequestToSpec` feels out of place at the moment)

File: packages/unmock-core/src/__tests__/serviceStore.test.ts [hunk 2]
```diff
@@ -1,65 +1,135 @@
...
+    expect(store.petstore).toThrow("has no defined paths");
   });
 
-  test("state class with basic call to services", () => {
-    const state = serviceStoreFactory(EmptyPathsServicePopulator);
-    state.petstore(); // Should pass
+  test("Store with empty paths", () => {
+    const store = serviceStoreFactory(EmptyPathsServicePopulator);
+    expect(store.petstore).toThrow("has no defined paths");
+    expect(store.petstore.get).toThrow("has no defined paths");
```
Comment (idantene) [inline packages/unmock-core/src/__tests__/serviceStore.test.ts]: These now throw because it doesn't make sense to set a state for a service that has no endpoints. Same for the changes below.

File: packages/unmock-core/src/__tests__/serviceStore.test.ts [hunk 3]
```diff
@@ -1,65 +1,135 @@
...
   });
 
-  test("state class with REST method calls to services", () => {
-    const state = serviceStoreFactory(EmptyPathsServicePopulator);
-    state.petstore.post(); // Should pass
+  test("Store with non-empty paths with non-matching method", () => {
+    const store = serviceStoreFactory(PetsPathsServicePopulator);
+    expect(store.petstore.post).toThrow("Can't find any endpoints with method");
```
Comment (idantene) [inline packages/unmock-core/src/__tests__/serviceStore.test.ts]: Same idea as above - what does it mean trying to set a state for "all paths with POST", where there are none? Hence it throws.

File: packages/unmock-core/src/service/interfaces.ts [hunk 4]
```diff
@@ -25,6 +25,18 @@ export interface IUnmockServiceState {
   [key: string]: UnmockServiceState;
 }
 
+export interface IService {
+  schema: OASSchema;
+  hasDefinedPaths: boolean;
+  findEndpoint(endpoint: string): string | undefined;
```
Comment (ksaaskil) [inline packages/unmock-core/src/service/interfaces.ts]: What does this do?

File: packages/unmock-core/src/service/service.ts [hunk 5]
```diff
@@ -1,37 +1,123 @@
...
-export class Service {
+
+export class Service implements IService {
   // Maintains a state for service
   // First level kv pairs: paths -> methods
   // Second level kv pairs: methods -> status codes
   // Third level kv pairs: status codes -> response template overrides
   // Fourth and beyond: template-specific.
-  // @ts-ignore
+  // @ts-ignore // ignored because it's currently only being read and not written
```
Comment (ksaaskil) [inline packages/unmock-core/src/service/service.ts]: Doesn't our TS config allow that :o Seems to me more like a thing linter would complain about.

File: packages/unmock-core/src/service/service.ts [hunk 6]
```diff
@@ -48,17 +134,62 @@ export class Service {
...
-  //     // make sure the combination exists
-  //   }
-  // }
+  private __complementPathWithRegexp(newPath: string, oldPath: string) {
+    // Complements oldPath by adding an unmock extension for regex matching the path.
+    const newPathRegex = XRegExp(`^${newPath}$`, "g");
+    this.oasSchema.paths[oldPath][UNMOCK_PATH_REGEX_KW] = newPathRegex;
+  }
+
+  private __updateSchemaPaths() {
```
Comment (ksaaskil) [inline packages/unmock-core/src/service/service.ts]: Minor thing, but why use underscore if it's already a `private`? Seems like Pythonism to me :)

File: packages/unmock-core/src/util.ts [hunk 7]
```diff
@@ -17,6 +17,7 @@ export const getAtLevel = (
   nestedObj: any,
   level: number,
   filterFn?: (key: string, value: any) => boolean,
+  noKeys: boolean = false,
```
Comment (ksaaskil) [inline packages/unmock-core/src/util.ts]: Brr, boolean argument :)

File: packages/unmock-core/src/service/service.ts [hunk 8]
```diff
@@ -48,17 +134,62 @@ export class Service {
...
+    Object.keys(this.schema.paths).forEach((path: string) => {
```
Comment (ksaaskil) [inline packages/unmock-core/src/service/service.ts]: Quite a massive thing inside a `forEach`, could this be somehow split into smaller pieces or include a docstring?

File: packages/unmock-core/src/service/interfaces.ts (hunk 4 above)
Comment (idantene) [reply packages/unmock-core/src/service/interfaces.ts]: Will add documentation (_copy, paste_ your doc style in #52 )

File: packages/unmock-core/src/service/service.ts (hunk 5 above)
Comment (idantene) [reply packages/unmock-core/src/service/service.ts]: It doesn't. I didn't want to take matters into my own hands as the state is temporarily not being used. Seems like a good protection IMO, linting-wise.

File: packages/unmock-core/src/service/service.ts (hunk 6 above)
Comment (idantene) [reply packages/unmock-core/src/service/service.ts]: Very true :laughing: Renaming

File: packages/unmock-core/src/service/service.ts (hunk 8 above)
Comment (idantene) [reply packages/unmock-core/src/service/service.ts]: I feel like it includes many docstrings, but I'll try and break it down further anyway. Thanks!

**Comment by idantene** [comment]
> Looks good to me! I'm a bit worried of `Service` becoming a hard-to-understand state machine, it could be unavoidable but still we should do what we can to avoid it. IMHO it would be ideal if one could wrap all state changes to e.g. a `state` member and one could reset the "state" with a dead-simple `this.state = new ServiceState()` or `this.state.reset()`.

At the moment there's nothing that deals with `state` directly, it's only being written.
There's a lot of OAS management though, making sure requests are valid, paths are found, methods are okay, etc -- I'm not sure that's avoidable.

**Comment by ksaaskil** [comment]
> > Looks good to me! I'm a bit worried of `Service` becoming a hard-to-understand state machine, it could be unavoidable but still we should do what we can to avoid it. IMHO it would be ideal if one could wrap all state changes to e.g. a `state` member and one could reset the "state" with a dead-simple `this.state = new ServiceState()` or `this.state.reset()`.
> 
> At the moment there's nothing that deals with `state` directly, it's only being written.
> There's a lot of OAS management though, making sure requests are valid, paths are found, methods are okay, etc -- I'm not sure that's avoidable.

Hmm, I don't understand how the OAS management requires `Service` to be changing its own state?

**Comment by idantene** [comment]
> Hmm, I don't understand how the OAS management requires `Service` to be changing its own state?

You're mixing these up :)
The `Service` class verifies `X` against the OAS, and maintains a state. Maintaining the state is trivial.
The verification against OAS is the heavy lifting the `Service` class actually does.

IMO it makes no sense to call `states.myService({...})` if `myService` has no endpoints. Similarly, it makes no sense to set `states.petstore("/pet/2", {...})`, if the spec doesn't specify that endpoint.
The latter fails early instead of silently accepting an invalid state for an endpoint that does not exist (however, `states.petstore("/pets/2", {...})` is fine, as the spec defines `/pets/{petId}` and describes what `petId` actually is).

**Comment by ksaaskil** [comment]
I agree 100%, I think we're speaking of different states here. I'm just saying `Service` should have as few changes to its internal members as possible.

**Comment by idantene** [comment]
Ah, I agree with that too. The only change happening to internal members is supplementing every path with an `x-unmock-path-regex`, upon creating a Service. It's also a sanity check for OAS.

Hopefully, once we join `loas` in, we can make some simplifying assumptions (e.g. every path parameter is indeed described under `parameters`), thereby removing many conditionals.
